src/main/java/bio/terra/common/retry/SamRetry.java

```diff
@@ -0,0 +1,194 @@
+package bio.terra.common.retry;
+
+import static java.time.Instant.now;
+
+import bio.terra.common.exception.BadRequestException;
+import bio.terra.common.exception.ConflictException;
+import bio.terra.common.exception.ErrorReportException;
+import bio.terra.common.exception.ForbiddenException;
+import bio.terra.common.exception.InternalServerErrorException;
+import bio.terra.common.exception.NotFoundException;
+import bio.terra.common.exception.UnauthorizedException;
+import com.fasterxml.jackson.core.JsonProcessingException;
+import com.fasterxml.jackson.databind.ObjectMapper;
+import com.google.api.client.http.HttpStatusCodes;
+import com.google.common.annotations.VisibleForTesting;
+import java.time.Duration;
+import java.time.Instant;
+import java.util.concurrent.TimeUnit;
+import org.apache.commons.lang3.StringUtils;
+import org.broadinstitute.dsde.workbench.client.sam.ApiException;
+import org.broadinstitute.dsde.workbench.client.sam.model.ErrorReport;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.springframework.http.HttpStatus;
+
+/**
+ * SamRetry encapsulates logic needed for retrying Sam API calls. All constants are hard-coded, as
+ * Sam failures are generally just short periods of down-time.
+ */
+public class SamRetry {
+
+  private static Logger logger = LoggerFactory.getLogger(SamRetry.class);
+
+  // The retry function starts with INITIAL_WAIT between retries, and doubles that until it
+  // reaches MAXIMUM_WAIT, after which all retries are MAXIMUM_WAIT apart.
+  private static final Duration MAXIMUM_WAIT = Duration.ofSeconds(30);
+  private static final Duration INITIAL_WAIT = Duration.ofSeconds(10);
+  private static final Duration OPERATION_TIMEOUT = Duration.ofSeconds(300);
+  private final Instant operationTimeout;
+
+  // How long to wait between retries.
+  private Duration retryDuration;
+
+  private static final ObjectMapper objectMapper = new ObjectMapper();
+
+  SamRetry() {
+    this.operationTimeout = now().plus(OPERATION_TIMEOUT);
+    this.retryDuration = INITIAL_WAIT;
+  }
+
+  @VisibleForTesting
+  SamRetry(Duration timeout) {
+    this.operationTimeout = now().plus(timeout);
+    this.retryDuration = INITIAL_WAIT;
+  }
+
+  @FunctionalInterface
+  public interface SamVoidFunction {
+    void apply() throws ApiException, InterruptedException;
+  }
+
+  @FunctionalInterface
+  public interface SamFunction<R> {
+    R apply() throws ApiException, InterruptedException;
+  }
+
+  public static <T> T retry(SamFunction<T> function) throws InterruptedException {
+    SamRetry samRetry = new SamRetry();
+    return samRetry.perform(function);
+  }
+
+  // Version of the retry function for testing so that we can override timeout
+  @VisibleForTesting
+  public static <T> T retry(SamFunction<T> function, Duration timeout) throws InterruptedException {
```

Comment:
nit: the `@VisibleForTesting` annotated methods can probably be package public instead of public if the tests are in the same package.